### PR TITLE
Fix pylint warnings, part 2

### PR DESCRIPTION
**Pull request by chmeliik**
Part 2 of pylint issue fixes.

For attributes defined outside `__init__`, I usually just defined them as `None`, though in some cases it seemed possible to turn them into basic local variables so I did that.

For global statements, I just made slight workarounds to get rid of the `global` statement, but `docker_mock.py` still heavily relies on global variables.

Also, `test_tasker.py` uses `docker.Client`, which no longer exists. I don't know where (or if) these tests are used with NOMOCK, so I just added a TODO about it.

Maintainers will complete the following section:
- [x] Commit messages are descriptive enough
- [x] "Signed-off-by:" line is present in each commit
- [x] Code coverage from testing does not decrease and new code is covered
- [x] JSON/YAML configuration changes are updated in the relevant schema
- [x] Changes to metadata also update the documentation for the metadata
- [x] Pull request includes link to an osbs-docs PR for user documentation updates

### CODE REVIEW

File: atomic_reactor/plugins/pre_pull_base_image.py [hunk 1]
```diff
@@ -71,12 +71,13 @@ def __init__(self, tasker, workflow, parent_registry=None, parent_registry_insec
...
             metadata.update(parent_images_digests)
 
+        self.manifest_list_cache = {}
+
     def run(self):
         """
         Pull parent images and retag them uniquely for this build.
         """
         build_json = get_build_json()
-        self.manifest_list_cache = {}
```
Comment (MartinBasti) [inline atomic_reactor/plugins/pre_pull_base_image.py]: Removing such initialization from run, may cause issues when we use the same instance for multiple runs. However I don't think that this is atomic-reactor case. Hopefully somebody will correct me if I'm wrong.

**Comment by MartinBasti** [comment]
Needs rebase

File: atomic_reactor/plugins/pre_pull_base_image.py (hunk 1 above)
Comment (MartinBasti) [reply atomic_reactor/plugins/pre_pull_base_image.py]: @twaugh @rcerven @athos-ribeiro

Comment (rcerven) [reply atomic_reactor/plugins/pre_pull_base_image.py]: yeah we always have just 1 plugin instance running

File: tests/docker_mock.py [hunk 2]
```diff
@@ -387,8 +385,7 @@ def remove_volume(volume_name):
              .and_raise(docker.errors.APIError, "xyz", response))
 
     if remember_images:
-        global mock_images
```
Comment (MartinBasti) [inline tests/docker_mock.py]: effectively this still works like global variable, it just bypassed warning and obfuscated code. I'm not sure if this worth fixing.

File: tests/test_tasker.py [hunk 3]
```diff
@@ -40,12 +40,12 @@
 def setup_module(module):
     if MOCK:
         return
-    d = docker.Client()
+    d = docker.Client()  # TODO: current version of python-docker does not have Client anymore
```
Comment (MartinBasti) [inline tests/test_tasker.py]: IMHO this needs investigation, we may need to freeze python-docker version in requirements.txt

File: tests/plugins/test_pulp.py [hunk 4]
```diff
@@ -45,7 +45,7 @@ class X(object):
     base_image = ImageName(repo="qwe", tag="asd")
 
 
-def prepare(check_repo_retval=0, existing_layers=[],
```
Comment (MartinBasti) [inline tests/plugins/test_pulp.py]: IMHO this changes behavior. `None` and `[]` have different meaning here.

You can use

```
_undefined = object()

def prepare(.., existing_layers=_undefined, ...):
    if existing_layers is _undefined:
       existing_layers = []
```

another way is, that all tests must be updated accordingly (preferred given the small number of tests using this function)

File: tests/plugins/test_rpmqa.py [hunk 5]
```diff
@@ -49,16 +49,20 @@ def mock_logs_empty(cid, **kwargs):
     return ''
 
 
-def mock_logs_retry(cid, cache={}, **kwargs):
```
Comment (MartinBasti) [inline tests/plugins/test_rpmqa.py]: Why this change in API? If needed it should be in separate commit. I don't see relation between fixing pylint warning and this change.

Comment (chmeliik) [reply tests/plugins/test_rpmqa.py]: IMO this was one of the rare cases where pylint actually reported a relevant warning. This method is not really usable more than once, because there is no good way to reset the cache. You would have to do something like this:
```python
retry_cache = {}
mock_logs_retry = partial(mock_logs_retry, cache=retry_cache)
# run a test
retry_cache.clear()
# run another test
retry_cache.clear()
# ...
```
The change essentially forces you to use clean cache every time you want to mock that method.
But if you think it should be in a separate commit that's ok with me.

File: tests/plugins/test_pulp.py (hunk 4 above)
Comment (chmeliik) [reply tests/plugins/test_pulp.py]: Ok, updated tests

File: tests/docker_mock.py (hunk 2 above)
Comment (chmeliik) [reply tests/docker_mock.py]: Ok, reverted and disabled the warning

**Comment by chmeliik** [comment]
Rebased on master

File: atomic_reactor/plugins/pre_pull_base_image.py (hunk 1 above)
Comment (chmeliik) [reply atomic_reactor/plugins/pre_pull_base_image.py]: Maybe the cache could just be cleared in `run()`, just to be sure?

File: tests/plugins/test_rpmqa.py (hunk 5 above)
Comment (MartinBasti) [reply tests/plugins/test_rpmqa.py]: Yes please, use separate commit with description you provided here :)

That makes git archaeology much easier.

File: atomic_reactor/plugins/pre_pull_base_image.py (hunk 1 above)
Comment (MartinBasti) [reply atomic_reactor/plugins/pre_pull_base_image.py]: Yes, cache invalidation is one of the 4 most common programming issues. Please just to be clear it in run.

**Comment by MartinBasti** [comment]
@chmeliik Any idea why coverage decreased here? https://coveralls.io/builds/23800932/source?filename=atomic_reactor/plugins/prepub_squash.py

**Comment by chmeliik** [comment]
@MartinBasti I found out it was because pylint doesn't know how to run test classes that have an `__init__` method. I also now clear cache when that plugin is run, and added a separate commit for the `mock_logs_retry` change.

File: tests/plugins/test_add_help.py [hunk 6]
```diff
@@ -26,14 +26,14 @@
 
 
 class Y(object):
-    pass
+    def __init__(self):
+        self.dockerfile_path = None
+        self.path = None
```
Comment (twaugh) [inline tests/plugins/test_add_help.py]: Alternatively, we could use `stubs.StubSource` instead of `Y`, and `stubs.StubInsideBuilder` instead of `X`.

File: tests/plugins/test_delete_from_registry.py [hunk 7]
```diff
@@ -39,14 +39,14 @@
 
 
 class Y(object):
-    pass
+    def __init__(self):
+        self.dockerfile_path = None
+        self.path = None
```
Comment (twaugh) [inline tests/plugins/test_delete_from_registry.py]: Alternatively, we could use `stubs.StubSource` instead of `Y`, and `stubs.StubInsideBuilder` instead of `X`.

File: tests/plugins/test_group_manifests.py [hunk 8]
```diff
@@ -302,14 +302,14 @@ def mock_registries(registries, config, schema_version='v2', foreign_layers=Fals
 
 
 class Y(object):
-    pass
+    def __init__(self):
+        self.dockerfile_path = None
+        self.path = None
```
Comment (twaugh) [inline tests/plugins/test_group_manifests.py]: Alternatively, we could use `stubs.StubSource` instead of `Y`, and `stubs.StubInsideBuilder` instead of `X`.

File: tests/plugins/test_tag_and_push.py [hunk 9]
```diff
@@ -83,14 +83,14 @@
 
 
 class Y(object):
-    pass
+    def __init__(self):
+        self.dockerfile_path = None
+        self.path = None
```
Comment (twaugh) [inline tests/plugins/test_tag_and_push.py]: Alternatively, we could use `stubs.StubSource` instead of `Y`, and `stubs.StubInsideBuilder` instead of `X`.

File: tests/plugins/test_tag_by_labels.py [hunk 10]
```diff
@@ -22,14 +22,14 @@
 
 
 class Y(object):
-    pass
+    def __init__(self):
+        self.dockerfile_path = None
+        self.path = None
```
Comment (twaugh) [inline tests/plugins/test_tag_by_labels.py]: Alternatively, we could use `stubs.StubSource` instead of `Y`, and `stubs.StubInsideBuilder` instead of `X`.

File: tests/test_tasker.py (hunk 3 above)
Comment (twaugh) [reply tests/test_tasker.py]: Should we try `APIClient` and fall back to `Client`, like we've done elsewhere?

**Comment by chmeliik** [comment]
@twaugh thanks, using `stubs.StubInsideBuilder` now. I felt like this change deserved a separate commit, as it is not really about fixing attributes defined outside init.

**Comment by twaugh** [comment]
Not sure what's up with Travis CI. Does it need a prod?

**Comment by chmeliik** [comment]
Pushed again to wake Travis up. Should I replace all those `X` classes with `StubInsideBuilder` as part of this PR? There is around 20 of them in various places.

**Comment by twaugh** [comment]
No need to scramble to replace `X`, just if we happen to need to touch them I'd rather we phased them out.

**Comment by twaugh** [comment]
The coverage decrease is only due to the `except ImportError` clauses handling modulemd not being available. I think we can ignore that.

**Comment by MartinBasti** [comment]
No release notes needed
